hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -3923,6 +3944,148 @@ DirectoryListing getListing(String src, byte[] startAfter,
     return dl;
   }
 
+  public byte[] getSrcPathsHash(String[] srcs) {
+    MessageDigest md;
+    try {
+      md = MessageDigest.getInstance("MD5");
```

Comment:
We should refine this method.

1. synchronize this method.
2. MessageDigest.getInstance() has big overhead so we should reuse it. (It's instantiates a new object each time)
3. After use, call MessageDigest#reset(). 

https://stackoverflow.com/questions/13802627/when-to-use-messagedigest-reset

Reply:
Good point. Will fix this.